### DISCUSSION TITLE
@urql/storage-rn and automatic cache updates

### DISCUSSION THREAD

**Discussion by nephix** [urql-graphql/urql#2381] [Q&A] [answered]
Hey everyone,

when using the `@urql/storage-rn` package as outlined [here](https://formidable.com/open-source/urql/docs/graphcache/offline/#react-native) I feel like urql is losing the ability to do automatic cache updates. I feel like this has something to do with the [sync first, async last](https://formidable.com/open-source/urql/docs/advanced/authoring-exchanges/#synchronous-first-asynchronous-last) approach but I can't seem to figure out a working order for my exchanges.

Is there anything obvious in the order that should be changed? I did install the mentioned NetInfo and AsyncStorage packages from Expo.

```
import { offlineExchange } from '@urql/exchange-graphcache';
import { makeAsyncStorage } from '@urql/storage-rn';
import type { Client, CombinedError } from 'urql';
import {
  createClient,
  Provider,
  dedupExchange,
  fetchExchange,
  errorExchange,
} from 'urql';

const cache = offlineExchange<GraphCacheConfig>({
    storage: makeAsyncStorage(),
});

createClient({
    url: API_URL,
    fetchOptions: () => ({
      headers: { Authorization: token ? `Bearer ${token}` : '' },
    }),
    exchanges: [
      dedupExchange,
      fetchExchange,
      errorExchange({
        onError: (error: CombinedError) => setError(error),
      }),
      cache,
    ],
}),
```

**Comment by JoviDeCroock** [comment, marked as answer]
```js
createClient({
    url: API_URL,
    fetchOptions: () => ({
      headers: { Authorization: token ? `Bearer ${token}` : '' },
    }),
    exchanges: [
      dedupExchange,
      cache,
      errorExchange({
        onError: (error: CombinedError) => setError(error),
      }),
      fetchExchange,
    ],
}),
```

The reason behind this order is in the request flow:

- we want to deduplicate operations
- we want to see if we have cached data for an operation
- we want to fetch if there is no cached operation

response flow:

- we check if there is an error
- we update the cache with updated or new data
- we release the duplication lock

The sync exchanges: dedup, cache and error
The async exchanges: fetch

The persisted storage is only used for the initial hydration of data to the memory and then updated in microtask delays so never interrupting the sync-cycle

In yours you do not use the cache anymore at all

**Reply by nephix** [reply]
Hmm when I try your order it doesn't seem to work as well. If I remove the `exchanges` field entirely then it works though.

> In yours you do not use the cache anymore at all

Oops sorry, I was doing some debugging and forgot it when copy/pasting

**Reply by JoviDeCroock** [reply]
Define not working, seems good for me... removing it means dedup-cache-fetch. Have you tried without the offline-storage if so did it work? If not maybe it's something to do with the version, if so maybe a bug in the rn adapter? I can't see how though as the data is only at hydration... If not works applies to data hydration that might be it though

**Reply by nephix** [reply]
Sorry for not being precise, not working meant: Queries that have been fetched before weren't updated through the cache.

> Have you tried without the offline-storage if so did it work?

Do you mean like:

```typescript
exchanges: [
    dedupExchange,
    // cache,
    errorExchange({
      onError: (error: CombinedError) => setError(error),
    }),
    fetchExchange,
],
```
?

That didn't update related queries either. Somehow removing the `exchanges` field works.

**Reply by nephix** [reply]
>The sync exchanges: dedup, cache and error

Is cache sync, even though it says `makeAsyncStorage`?

**Reply by JoviDeCroock** [reply]
As my explanation there is a big difference between data-hydration and memory retrieval. The cache functions as normal the difference being a write to async storage within microticks (does not stop sync flows) and an async initial read to hydrate the cache.

The reason why it works without exchanges is also detailed there

**Reply by JoviDeCroock** [reply]
An example: https://codesandbox.io/s/wizardly-panna-y2moe?file=/src/index.js updateTodo can also just work automatically adding has a manual updater

**Reply by nephix** [reply]
Is it possible that it has something to do with a field of a mutation response being nullable and therefore not having a `__typename`?

**Reply by JoviDeCroock** [reply]
I mean 😅 that is very vague... but yes generally automatic updates won't do anything without data

**Reply by nephix** [reply]
It has data, but the data is nullable so it doesn't _always_ have a `__typename` 😉 . But still strange that it works without defining the exchanges manually 🤔
